### PR TITLE
HDDS-9944. NSSummary commands should close OzoneClient.

### PR DESCRIPTION
**Pull request by devmadhuu**
## What changes were proposed in this pull request?
This PR fixes the Ozone Client connection leak by closing all opened client connections in NSSummaryAdmin.java

NSSummaryAdmin commands leaves the Ozone client connections open which is not a good practice and may consume resources. So. this PR fixes that by closing all opened Ozone client connections.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9944

## How was this patch tested?

This patch was tested using existing integration test `org.apache.hadoop.ozone.shell.TestNSSummaryAdmin`

### CODE REVIEW

**Comment by devmadhuu** [comment]
@adoroszlai Pls review.

**Review by adoroszlai — COMMENTED** [review]
Thanks @devmadhuu for working on this.

There is a second item in the task:

> it should reuse client (or even bucket after lookup) for all checks

All `namespace` subcommands use these methods as the following:

https://github.com/apache/ozone/blob/2616f8e352285846f0d37ea9b7d9523d6762126d/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/DiskUsageSubCommand.java#L109-L112

I don't think we need to use two separate clients for checking properties of the same bucket.  The two methods can be combined.

Also, `isFileSystemOptimizedBucket` seems to be unused.

**Comment by devmadhuu** [comment]
> Thanks @devmadhuu for working on this.
> 
> There is a second item in the task:
> 
> > it should reuse client (or even bucket after lookup) for all checks
> 
> All `namespace` subcommands use these methods as the following:
> 
> https://github.com/apache/ozone/blob/2616f8e352285846f0d37ea9b7d9523d6762126d/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/DiskUsageSubCommand.java#L109-L112
> 
> I don't think we need to use two separate clients for checking properties of the same bucket. The two methods can be combined.
> 
> Also, `isFileSystemOptimizedBucket` seems to be unused.

Thanks @adoroszlai for your review. I have implemented 2nd part of this JIRA :
   --- NSSummaryAdmin all sub-commands should reuse ozone client (or even bucket after lookup) for all checks

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java [hunk 1]
```diff
@@ -86,86 +86,52 @@ public Class<?> getParentType() {
...
       System.out.println(
           "Bucket layout couldn't be verified for path: " + ofsPath +
               ". Exception: " + e);
-      return false;
+      bucket = null;
+    } finally {
+      ozoneClient.close();
     }
+    boolean isOBSBucket = bucket != null ? isObjectStoreBucket(bucket, objectStore) : false;
+    return isOBSBucket || !Objects.nonNull(bucket);
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java:134]: `!Objects.nonNull(bucket)` is a bit hard to decipher, please simplify.

Please move these inside the `try` block, and use `try-with-resources`.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java [hunk 2]
```diff
@@ -86,86 +86,52 @@ public Class<?> getParentType() {
...
    * Return false if path is root, just a volume or invalid.
+   * Returns false if bucket is part of path but not a OBS bucket.
    * @param path
-   * @return true if the bucket
-   * is not part of the given path.
+   * @return true if bucket is OBS bucket or not part of provided path.
    * @throws IOException
    */
-  public boolean bucketIsPresentInThePath(String path) throws IOException {
+  public boolean isOBSBucketOrNotPresentInPath(String path) throws IOException {
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java:114]: Let's try to find a better name.  Describe the intention, not the specific condition.

**Review by adoroszlai — COMMENTED** [review]
Thanks @devmadhuu for updating the patch with re-use of client.  Some minor items suggested.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java (hunk 1 above)
Comment (devmadhuu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java]: > `!Objects.nonNull(bucket)` is a bit hard to decipher, please simplify.
> 
> Please move these inside the `try` block, and use `try-with-resources`.

Ok, I have tried to simplify now.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java (hunk 2 above)
Comment (devmadhuu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java]: > Let's try to find a better name. Describe the intention, not the specific condition.

Ok, tried to give a better name.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java [hunk 3]
```diff
@@ -86,86 +85,47 @@ public Class<?> getParentType() {
...
-
-      return Objects.nonNull(bucket);
+      return isObjectStoreBucket(bucket, objectStore);
     } catch (IOException e) {
       System.out.println(
           "Bucket layout couldn't be verified for path: " + ofsPath +
               ". Exception: " + e);
-      return false;
     }
+    return false;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java:128]: `true` should be returned if bucket is not found.

```
Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 20.55 s <<< FAILURE! -- in org.apache.hadoop.ozone.shell.TestNSSummaryAdmin
org.apache.hadoop.ozone.shell.TestNSSummaryAdmin.testNSSummaryCLIRoot -- Time elapsed: 0.344 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:183)
	at org.apache.hadoop.ozone.shell.TestNSSummaryAdmin.testNSSummaryCLIRoot(TestNSSummaryAdmin.java:119)
```

https://github.com/devmadhuu/ozone/actions/runs/7579739079/job/20644851580#step:6:141

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java [hunk 4]
```diff
@@ -86,86 +85,47 @@ public Class<?> getParentType() {
...
+  public boolean isNotValidBucketOrOBSBucket(String path) {
     OFSPath ofsPath = new OFSPath(path,
         OzoneConfiguration.of(getOzoneConfig()));
-
-    OzoneClient ozoneClient = OzoneClientFactory.getRpcClient(getOzoneConfig());
-    ObjectStore objectStore = ozoneClient.getObjectStore();
-
-    try {
-      OzoneBucket bucket = objectStore.getVolume(ofsPath.getVolumeName())
+    OzoneBucket bucket;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java:116]: nit: Can be joined with the assignment inside the `try` block.

**Review by adoroszlai — COMMENTED** [review]
Thanks @devmadhuu for updating the patch.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java (hunk 4 above)
Comment (devmadhuu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java]: Done.

**Comment by adoroszlai** [comment]
@ArafatKhan2198 please review

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java [hunk 5]
```diff
@@ -86,86 +85,46 @@ public Class<?> getParentType() {
...
    * Return false if path is root, just a volume or invalid.
+   * Returns false if bucket is part of path but not a OBS bucket.
    * @param path
-   * @return true if the bucket
-   * is not part of the given path.
+   * @return true if bucket is OBS bucket or not part of provided path.
    * @throws IOException
    */
-  public boolean bucketIsPresentInThePath(String path) throws IOException {
+  public boolean isNotValidBucketOrOBSBucket(String path) {
```
Comment (xBis7) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java:113]: I can see that for every command, this check is preceded by a check to whether the path exists. Therefore, there is no issue of a NPException.

**Review by xBis7 — APPROVED** [review]
@devmadhuu Thanks for the patch. I've tested it locally. LGTM!

**Review by ArafatKhan2198 — APPROVED** [review]
Thank you for the modifications, @devmadhuu. I was just curious about why we are directly utilizing `println()` instead of employing logging.

```
System.out.println(
    "Bucket layout couldn't be verified for path: " + ofsPath +
        ". Exception: " + e);
```

**Comment by adoroszlai** [comment]
Thanks @devmadhuu for the fix, @ArafatKhan2198, @xBis7 for the review.

**Comment by devmadhuu** [comment]
> Thank you for the modifications, @devmadhuu. I was just curious about why we are directly utilizing `println()` instead of employing logging. I believe even for a CLI command it usually does log the message on the console right?
> 
> ```
> System.out.println(
>     "Bucket layout couldn't be verified for path: " + ofsPath +
>         ". Exception: " + e);
> ```

Thanks @ArafatKhan2198 for review. These are CLI based interface, so need to print on console.

**Pull request merged**